vpr/src/base/ShowSetup.cpp

```diff
@@ -61,16 +63,18 @@ void ShowSetup(const t_vpr_setup& vpr_setup) {
     }
 }
 
-void printClusteredNetlistStats() {
+void printClusteredNetlistStats(std::string block_usage_filename) {
     auto& device_ctx = g_vpr_ctx.device();
     auto& cluster_ctx = g_vpr_ctx.clustering();
 
```

Comment:
I would recommend creating a struct called something like `ClusteredNetlistStats` and four functions;

```cpp
ClusteredNetlistStats& createClusteredNetlistStats() { ... }

void writeClusteredNetlistStatsHuman(const ClusteredNetlistStats& stats, std::ostream output) { ... }
void writeClusteredNetlistStatsJSON(const ClusteredNetlistStats& stats, std::ostream output) { ... }
void writeClusteredNetlistStatsXML(const ClusteredNetlistStats& stats, std::ostream output) { ... }

void writeClusteredNetlistStats(std::string block_usage_filename) {
    auto& stats = createClusteredNetlistStats(g_vpr_ctx);
    // Print out the human readable version to stdout
    if (!g_vpr_ctx.quiet) {
        writeClusteredNetlistStatsHuman(stats, std::stdout);
    }

    // Save a machine readable version
    if (block_usage_filename.size() > 0) {
        std::fstream fp;
        fp.open(block_usage_filename, std::fstream::out | std::fstream::trunc);

        if (vtr::check_file_name_extension(block_usage_filename.c_str(), ".json")) {
            writeClusteredNetlistStatsJSON(stats, fp);
        } else if (vtr::check_file_name_extension(block_usage_filename.c_str(), ".xml")) {
            writeClusteredNetlistStatsXML(stats, fp);
        } else if (vtr::check_file_name_extension(block_usage_filename.c_str(), ".txt")) {
            writeClusteredNetlistStatsHuman(stats, fp);
        } else {
            VPR_FATAL_ERROR(
                VPR_ERROR_PACK,
                "Unknown extension on output %s",
                block_usage_filename.c_str());
        }
       fp.close();
    }
}
```

or maybe;
```cpp

struct ClusteredNetlistStats {
    private:
        void writeHuman(std::ostream output) const { ... }
        void writeJSON(std::ostream output) const { ... }
        void writeXML(std::ostream output) const { ... }

    public:
        ClusteredNetlistStats(g_vpr_ctx);

        enum OutputFormat {
            HumanReadable,
            JSON,
            XML
        };

        void write(OutputFormat fmt, std::ostream output) const {
            switch (fmt) {
                case HumanReadable:
                    return writeHuman(output);
                case JSON:
                    return writeJSON(output);
                case XML:
                    return writeXML(output);
            }
        }
}

void writeClusteredNetlistStats(std::string block_usage_filename) {
    const auto stats = ClusteredNetlistStats(g_vpr_ctx);

    // Print out the human readable version to stdout
    if (!g_vpr_ctx.quiet) {
        stats.write(ClusteredNetlistStatsHuman::OutputFormat::HumanReadable, std::stdout);
    }

    if (block_usage_filename.size() > 0) {
        ClusteredNetlistStats::OutputFormat fmt;
        if (vtr::check_file_name_extension(block_usage_filename.c_str(), ".json")) {
            fmt = ClusteredNetlistStats::OutputFormat::JSON;
        } else if (vtr::check_file_name_extension(block_usage_filename.c_str(), ".xml")) {
            fmt = ClusteredNetlistStats::OutputFormat::XML;
        } else if (vtr::check_file_name_extension(block_usage_filename.c_str(), ".txt")) {
            fmt = ClusteredNetlistStats::OutputFormat::HumanReadable;
        } else {
            VPR_FATAL_ERROR(
                VPR_ERROR_PACK,
                "Unknown extension on output %s",
                block_usage_filename.c_str());
        }
        std::fstream fp;
        fp.open(block_usage_filename, std::fstream::out | std::fstream::trunc);
        stats.write(fmt, fp);
        fp.close();
    }
}
```